TypeCobol.LanguageServer/Workspace.cs

```diff
@@ -951,6 +952,209 @@ private void FinalCompilationStepCompleted(object cUnit, ProgramClassEvent progr
+                    {
+                        if (dataDefinition.Picture != null)
+                        {
+                            row.Append(string.Format(PIC, dataDefinition.Picture.Value));
+                        }
+                        string usage = ((DataDescriptionEntry) dataDefinition.CodeElement).Usage?.Token.Text;
+                        if (!string.IsNullOrWhiteSpace(usage))
+                        {
+                            AppendSpaceIfNeeded();
+                            row.Append(usage);
+                        }
+
+                        if (row.Length == initialRowLength && dataDefinition.ChildrenCount > 0)
+                        {
+                            row.Append(GROUP);
+                        }
+                    }
+
+                    if (dataDefinition.IsTableOccurence)
+                    {
+                        AppendSpaceIfNeeded();
+                        row.Append(string.Format(OCCURS, dataDefinition.MaxOccurencesCount));
+                    }
+
+                    row.Append(separator);
+
+                    void AppendSpaceIfNeeded()
+                    {
+                        if (row.Length != initialRowLength)
+                        {
+                            row.Append(SPACE);
+                        }
+                    }
+                }
+
+                static object GetEnd(long start, long length)
+                {
+                    var end = start + length - 1;
+                    return (end > 0) ? end : UNDEFINED;
+                }
+            }
+        }
+
+        private List<Tuple<int, DataDefinition, int>> CollectDataLayoutNodes(CompilationUnit compilationUnit)
+        {
+            var dataLayoutNodes = new List<Tuple<int, DataDefinition, int>>();
+            Node dataDivision = compilationUnit.TemporaryProgramClassDocumentSnapshot.Root.MainProgram.GetChildren<DataDivision>().FirstOrDefault();
+            if (dataDivision != null)
+            {
+                // Consider data declared in the Working storage
+                var workingStorage = dataDivision.GetChildren<WorkingStorageSection>().FirstOrDefault();
+                if (workingStorage != null)
+                {
+                    CollectDataLayoutNodes(0, workingStorage, 0);
+                }
+
+                // Consider also data declared in the Local storage
+                var localStorage = dataDivision.GetChildren<LocalStorageSection>().FirstOrDefault();
+                if (localStorage != null)
+                {
+                    CollectDataLayoutNodes(0, localStorage, 0);
+                }
+            }
+
+            return dataLayoutNodes;
+
+            void CollectDataLayoutNodes(int nodeLevel, Node node, int occursDimension)
+            {
+                foreach (var child in node.Children)
+                {
+                    if (child is DataDefinition childDefinition)
```

Comment:
This can be simplified using

```csharp
foreach (var childDefinition in node.Children.OfType<DataDefinition>())
{
    [...]
}
```

But this is for your information, you can keep code like that.

Reply:
Ok, good to know, thanks.